diff -u a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -3,4 +3,8 @@
-set(CMAKE_TOOLCHAIN_FILE "$ENV{VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake" CACHE STRING "")
-set(VCPKG_TARGET_TRIPLET "x64-windows-static" CACHE STRING "")
+if(DEFINED ENV{VCPKG_ROOT} AND NOT DEFINED CMAKE_TOOLCHAIN_FILE)
+	set(CMAKE_TOOLCHAIN_FILE "$ENV{VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake" CACHE STRING "")
+endif()
+if(NOT DEFINED VCPKG_TARGET_TRIPLET)
+	set(VCPKG_TARGET_TRIPLET "x64-windows-static-md" CACHE STRING "")
+endif()
 
@@ -43,3 +47,4 @@
 	add_compile_definitions(
 		_UNICODE
+		EXCLUSIVE_SKYRIM_AE=1
 	)
@@ -106,4 +111,5 @@
 	set(LINK_OPTIONS_RELEASE
 		/INCREMENTAL:NO	# Link Incrementally
 		/OPT:REF	# Optimizations (eliminate functions/data never referenced)
 		/OPT:ICF	# Optimizations (perform identical COMDAT folding)
+		/DEBUG:FULL
 	)
@@ -114,1 +120,1 @@
-add_subdirectory(extern/CommonLibSSE CommonLibSSE)
+find_package(CommonLibSSE CONFIG REQUIRED)
@@ -147,7 +153,6 @@
 set_property(
 	TARGET
 		${PROJECT_NAME}
-		CommonLibSSE
 	PROPERTY
-		MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>"
+		MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>DLL"
 )
@@ -157,1 +162,1 @@
-		cxx_std_17
+		cxx_std_20
diff -u a/src/PCH.h b/src/PCH.h
--- a/src/PCH.h
+++ b/src/PCH.h
@@ -5,4 +5,4 @@
-#include <Windows.h>
 #include "RE/Skyrim.h"
 #include "SKSE/SKSE.h"
 #include "xbyak/xbyak.h"
+#include <Windows.h>
@@ -25,0 +25,2 @@
+#include <span>
+
@@ -30,0 +32,1 @@
+#define SKSEAPI __cdecl
@@ -34,1 +37,1 @@
-    using nonstd::span;
+    using std::span;
--- a/src/main.cpp
+++ b/src/main.cpp
@@ -1,3 +1,14 @@
+SKSE_PLUGIN_VERSION = []() {
+    SKSE::PluginVersionData v;
+    v.PluginVersion(REL::Version{ 1, 0, 0, 0 });
+    v.PluginName("SinkOrSwim");
+    v.AuthorName("Loki");
+    v.UsesAddressLibrary();
+    v.UsesUpdatedStructs();
+    v.UsesNoStructs();
+    return v;
+}();
+
 extern "C" DLLEXPORT bool SKSEAPI SKSEPlugin_Query(const SKSE::QueryInterface * a_skse, SKSE::PluginInfo * a_info)
 {
 #ifndef NDEBUG
@@ -56,16 +67,101 @@
         return result;
 
     }
+    static inline bool ActorHasPowerArmor(RE::Actor* refr)
+    {
+        if (!refr) return false;
+
+        std::vector<RE::BGSBipedObjectForm::BipedObjectSlot> bodySlots = {
+            RE::BGSBipedObjectForm::BipedObjectSlot::kModChestPrimary,
+            RE::BGSBipedObjectForm::BipedObjectSlot::kBody,
+            RE::BGSBipedObjectForm::BipedObjectSlot::kModChestSecondary
+        };
+
+        std::vector<RE::BGSBipedObjectForm::BipedObjectSlot> handSlots = {
+            RE::BGSBipedObjectForm::BipedObjectSlot::kHands,
+            RE::BGSBipedObjectForm::BipedObjectSlot::kForearms,
+            RE::BGSBipedObjectForm::BipedObjectSlot::kModArmLeft,
+            RE::BGSBipedObjectForm::BipedObjectSlot::kModArmRight
+        };
+
+        std::vector<RE::BGSBipedObjectForm::BipedObjectSlot> headSlots = {
+            RE::BGSBipedObjectForm::BipedObjectSlot::kCirclet,
+            RE::BGSBipedObjectForm::BipedObjectSlot::kHair,
+            RE::BGSBipedObjectForm::BipedObjectSlot::kLongHair,
+            RE::BGSBipedObjectForm::BipedObjectSlot::kHead
+        };
+
+        std::vector<RE::BGSBipedObjectForm::BipedObjectSlot> feetSlots = {
+            RE::BGSBipedObjectForm::BipedObjectSlot::kFeet,
+            RE::BGSBipedObjectForm::BipedObjectSlot::kCalves,
+            RE::BGSBipedObjectForm::BipedObjectSlot::kModLegLeft,
+            RE::BGSBipedObjectForm::BipedObjectSlot::kModLegRight
+        };
+
+        auto hasHeavyBody = false;
+        for (auto slot : bodySlots) {
+            auto armor = refr->GetWornArmor(slot);
+            if (armor && armor->GetArmorType() == RE::BGSBipedObjectForm::ArmorType::kHeavyArmor) {
+                hasHeavyBody = true;
+                break;
+            }
+        }
+
+        auto hasHeavyHands = false;
+        for (auto slot : handSlots) {
+            auto armor = refr->GetWornArmor(slot);
+            if (armor && armor->GetArmorType() == RE::BGSBipedObjectForm::ArmorType::kHeavyArmor) {
+                hasHeavyHands = true;
+                break;
+            }
+        }
+
+        auto hasHeavyHead = false;
+        for (auto slot : headSlots) {
+            auto armor = refr->GetWornArmor(slot);
+            if (armor && armor->GetArmorType() == RE::BGSBipedObjectForm::ArmorType::kHeavyArmor) {
+                hasHeavyHead = true;
+                break;
+            }
+        }
+
+        auto hasHeavyFeet = false;
+        for (auto slot : feetSlots) {
+            auto armor = refr->GetWornArmor(slot);
+            if (armor && armor->GetArmorType() == RE::BGSBipedObjectForm::ArmorType::kHeavyArmor) {
+                hasHeavyFeet = true;
+                break;
+            }
+        }
+        return hasHeavyBody && hasHeavyHands && hasHeavyHead && hasHeavyFeet;
+    }
+
     static void InstallSwimmingHook()
     {
-        REL::Relocation<std::uintptr_t> target{ REL::ID(36357), 0x6ED };  // actor_update code insertion point
-        REL::Relocation<std::uintptr_t> isSwimmingVariable{ REL::ID(241932) };
+        REL::Relocation<std::uintptr_t> target{ REL::RelocationID(36357, 37348), REL::VariantOffset(0x6ED, 0x69D, 0) };  // actor_update code insertion point
         Loki_SinkOrSwim L_SOS;
 
         struct Patch : Xbyak::CodeGenerator
         {
-            Patch(std::uintptr_t a_variable, std::uintptr_t a_target)
+            Patch(std::uintptr_t a_target)
             {
+#if defined(EXCLUSIVE_SKYRIM_AE)
+                Xbyak::Label _OurReturn;
+                Xbyak::Label retLabel;
+
+                setae(sil);
+                mov(rcx, (uintptr_t)&isHeavy);
+                cmp(byte[rcx], 1);
+                jne(retLabel);
+                mov(sil, 0);   // set kIsSwimming to 0 if our own isHeavy is set to TRUE
+
+                L(retLabel);
+                mov(ptr[rbp + 0x1D0], sil);
+                jmp(ptr[rip + _OurReturn]);
+
+                L(_OurReturn);
+                dq(a_target + 0xB);  // 0x69D + 0xB = 0x6A8
+#else
                 Xbyak::Label _OurReturn;
                 Xbyak::Label retLabel;
                 Xbyak::Label isSwimmingAddr;
@@ -82,26 +178,33 @@
                 jmp(ptr[rip + _OurReturn]); // we do not touch the drowning flag or this address, but we overwrite
                                             // part of the code so we need to bring it into our injection
                 L(isSwimmingAddr);          // all of this happens directly AFTER GetSubmergedLevel()
-                dq(a_variable);
+                dq(REL::ID(241932).address());
 
                 L(_OurReturn);
                 dq(a_target + 0xB);  // 0x6F8 - 0x6ED
-            }
-        };
-
-        Patch patch(isSwimmingVariable.address(), target.address());
+#endif
+            }
+        };
+
+        Patch patch(target.address());
         patch.ready();
 
         auto& trampoline = SKSE::GetTrampoline();
-        trampoline.write_branch<6>(target.address(), L_SOS.CodeAllocation(patch, &trampoline));
+        trampoline.write_branch<6>(target.address(), L_SOS.CodeAllocation(patch, &trampoline), true);
+
+#if defined(EXCLUSIVE_SKYRIM_AE)
+        // Cleanly fill remaining 5 bytes of the 11-byte overwritten instruction block with NOPs
+        std::uint8_t nops[5] = { 0x90, 0x90, 0x90, 0x90, 0x90 };
+        REL::safe_write(target.address() + 6, nops, sizeof(nops));
+#endif
     }
 
     static void InstallWaterHook()
     {
-        REL::Relocation<std::uintptr_t> ActorUpdate{ REL::ID(36357) };
+        REL::Relocation<std::uintptr_t> ActorUpdate{ REL::RelocationID(36357, 37348) };
 
         auto& trampoline = SKSE::GetTrampoline();
-        _GetSubmergeLevel = trampoline.write_call<5>(ActorUpdate.address() + 0x6D3, GetSubmergeLevel);
+        _GetSubmergeLevel = trampoline.write_call<5>(ActorUpdate.address() + REL::VariantOffset(0x6D3, 0x683, 0).offset(), GetSubmergeLevel);
     }
 
 private:
@@ -127,29 +230,36 @@
 
         auto activeEffects = a_this->GetActiveEffectList();
         isHeavy = FALSE;  // set to false on logic start
+
+        bool isHeavyArmor = ActorHasPowerArmor(a_this);
+        if (!isHeavyArmor && activeEffects && WaterSlowdownSink) {
+            for (auto& ae : *activeEffects) {
+                if (ae && ae->spell == WaterSlowdownSink) {
+                    if (ae->flags.none(RE::ActiveEffect::Flag::kInactive) && ae->flags.none(RE::ActiveEffect::Flag::kDispelled)) {
+                        isHeavyArmor = true;
+                        break;
+                    }
+                }
+            }
+        }
+
         if (submergedLevel >= 0.69) {
             a_this->RemoveSpell(WaterSlowdownBeeg);
             a_this->RemoveSpell(WaterSlowdownSmol);
             a_this->AddSpell(WaterSlowdownSwim);
             a_this->AddSpell(WaterSlowdownSink);
 
-            if (activeEffects) {
-                for (auto& ae : *activeEffects) {
-                    if (ae && ae->spell == WaterSlowdownSink) {
-                        if (ae->flags.none(RE::ActiveEffect::Flag::kInactive) && ae->flags.none(RE::ActiveEffect::Flag::kDispelled)) {
-                            if (!a_this->GetCharController()) {
-                                logger::info("Invalid CharController ptr, skipping gravity code.");
-                            } else {
-                                a_this->GetCharController()->gravity = 0.20f;    // set gravity so we "float" when submerged, dont let it reset
-                                isHeavy = TRUE;
-                                if (!a_this->pad11) {  // we only need this to run ONCE when meeting the condition
-                                    const RE::hkVector4 hkv = { -1.00f, -1.00f, -1.00f, -1.00f };
-                                    a_this->GetCharController()->SetLinearVelocityImpl(hkv);
-                                    a_this->pad11 = TRUE;  // this is a (presumably) unused variable that i am putting to use
-                                    goto JustFuckingLeave;
-                                }
-                            }
-                        }
+            if (isHeavyArmor) {
+                if (!a_this->GetCharController()) {
+                    logger::info("Invalid CharController ptr, skipping gravity code.");
+                } else {
+                    a_this->GetCharController()->gravity = 0.20f;    // set gravity so we "float" when submerged, dont let it reset
+                    isHeavy = TRUE;
+                    if (!a_this->pad11) {  // we only need this to run ONCE when meeting the condition
+                        const RE::hkVector4 hkv = { -1.00f, -1.00f, -1.00f, -1.00f };
+                        a_this->GetCharController()->SetLinearVelocityImpl(hkv);
+                        a_this->pad11 = TRUE;  // this is a (presumably) unused variable that i am putting to use
+                        goto JustFuckingLeave;
                     }
                 }
             }
@@ -186,6 +296,15 @@
             };
         }
     JustFuckingLeave:
+        if (a_this->IsPlayerRef()) {
+            static bool s_lastHeavy = false;
+            static float s_lastSubmerged = -1.0f;
+            if (s_lastHeavy != isHeavy || std::abs(submergedLevel - s_lastSubmerged) > 0.25f) {
+                logger::info("Player submergedLevel={:.2f}, isHeavy={}, pad11={}", submergedLevel, isHeavy, (bool)a_this->pad11);
+                s_lastHeavy = isHeavy;
+                s_lastSubmerged = submergedLevel;
+            }
+        }
         return submergedLevel;
     };
     static inline REL::Relocation<decltype(GetSubmergeLevel)> _GetSubmergeLevel;
@@ -203,4 +322,4 @@
     Loki_SinkOrSwim::InstallSwimmingHook();
 
     return true;
-}
+}